### PR TITLE
Issue 311

### PR DESCRIPTION
**Pull request by rsk2**
For #311

### CODE REVIEW

File: pytition/petition/templates/petition/edit_petition.html [hunk 1]
```diff
@@ -90,16 +90,76 @@
...
+            </div>
+        </div>
+    </div>
+    <div class="form-group row">
+        <label for="id_has_tumblr_share_button" class="col-sm-12 col-md-3 col-lg-2 col-form-label">{% trans "Enable Tumblr share button?" %}</label>
+        <div class="col-sm-12 col-md-9 col-lg-10">
+            <div class="custom-control custom-switch">
+                <input type="checkbox" name="has_tumblr_share_button"  class="is-valid custom-control-input" id="id_has_tumblr_share_button"
+                        {% if petition.has_tumblr_share_button %} checked {% endif %}>
+                <label class="custom-control-label" for="id_has_tumblr_share_button">{% trans "Enable Tumblr share buttons?" %}</label>
```
Comment (fallen) [inline pytition/petition/templates/petition/edit_petition.html:119]: `button` without `s`

File: pytition/petition/templates/petition/petition_detail.html [hunk 2]
```diff
@@ -51,7 +51,7 @@
 
 {% block extracss %}
 <link href="{% static "css/petition.css" %}" rel="stylesheet" type="text/css">
-{% if petition.has_share_buttons %}
+{% if petition.has_email_share_button or petition.has_facebook_share_button or petition.has_tumblr_share_button or petition.has_linkedin_share_button or petition.has_twitter_share_button or petition.has_mastodon_share_button or petition.has_whatsapp_share_button %}
```
Comment (fallen) [inline pytition/petition/templates/petition/petition_detail.html:54]: you could add a `has_any_share_button()` method in the `Petition` class that does this boolean computation, just for the sake of keeping the template easy to read.

File: pytition/petition/templates/petition/petition_detail.html [hunk 3]
```diff
@@ -262,77 +262,91 @@ <h2 class="text-danger">{% trans "This petition is moderated. You only see it be
...
-                          </span>
-                          <span class="rrssb-text">Whatsapp</span>
-                        </a>
-                      </li>
-                    </ul>
-                  </div>
-              {% endif %}
+                {% if petition.has_email_share_button or petition.has_facebook_share_button or petition.has_tumblr_share_button or petition.has_linkedin_share_button or petition.has_twitter_share_button or petition.has_mastodon_share_button or petition.has_whatsapp_share_button %}
```
Comment (fallen) [inline pytition/petition/templates/petition/petition_detail.html:265]: Same remark here about adding a `has_any_share_button()` method to `Petition` class.

File: pytition/petition/templates/petition/petition_detail.html [hunk 4]
```diff
@@ -357,7 +371,7 @@ <h2 class="text-danger">{% trans "This petition is moderated. You only see it be
 
 {% block extrajs %}
 <script type="text/javascript" src="{% static "js/petition.js" %}"></script>
-{% if petition.has_share_buttons %}
+{% if petition.has_email_share_button or petition.has_facebook_share_button or petition.has_tumblr_share_button or petition.has_linkedin_share_button or petition.has_twitter_share_button or petition.has_mastodon_share_button or petition.has_whatsapp_share_button %}
```
Comment (fallen) [inline pytition/petition/templates/petition/petition_detail.html:374]: Same remark here.

**Review by fallen — COMMENTED** [review]
Awesome! 
This looks good, just a few remarks in the review :)
Also, could you add/update the testsuite to test for those fields?
Thanks a lot!

**Comment by fallen** [comment]
For the unit test you can either add a new test in `pytition/petition/tests/tests_EditPetitionView.py` or just modify `test_edit_petition_POST_social_network_form()`.
I would see it like this:
* Testing that submitting the form with one of the `has_xxx_share_button` set and that the Petition object then has the correct property value
* Then unset it and submit the form again and check that the property is unset
* Do this for each new property.
* You can kind of copy paste those tests in `pytition/petition/tests/tests_PetitionTemplateViews.py` and just modify the name of the view to submit the `edit_template` form instead of the `edit_petition` so that the new properties get tested for the `Template` model class as well.

**Comment by rsk2** [comment]
> For the unit test you can either add a new test in `pytition/petition/tests/tests_EditPetitionView.py` or just modify `test_edit_petition_POST_social_network_form()`. I would see it like this:

OK. I am planning to add new test in the file `pytition/petition/tests/tests_EditPetitionView.py`

>     * Testing that submitting the form with one of the `has_xxx_share_button` set and that the Petition object then has the correct property value

@fallen can you please give me an example test for one of the `has_xxx_share_button`?

**Comment by fallen** [comment]
> @fallen can you please give me an example test for one of the `has_xxx_share_button`?

Sure, here you go:

```
    def test_edit_petition_email_share_button(self):
        julia = self.login('julia')
        org = Organization.objects.get(name='RAP')
        share_button_form_data = {
            'social_network_form_submitted': 'yes',
            'has_email_share_button': True,
        }

        # For an org owned petition
        p = Petition.objects.create(title="My petition", org=org)

        # By default, a new petition has no share button
        self.assertFalse(p.has_email_share_button)

        # Now let's enable the email share button
        response = self.client.post(reverse("edit_petition", args=[p.id]),
                                    share_button_form_data)
        self.assertEqual(response.status_code, 200)
        p.refresh_from_db()
        self.assertTemplateUsed(response, "petition/edit_petition.html")
        self.assertEquals(response.context['social_network_form'].is_valid(), True)
        self.assertEquals(response.context['social_network_form'].is_bound, True)
        self.assertEquals(response.context['content_form_submitted'], False)
        self.assertEquals(response.context['email_form_submitted'], False)
        self.assertEquals(response.context['social_network_form_submitted'], True)
        self.assertEquals(response.context['newsletter_form_submitted'], False)
        self.assertTrue(p.has_email_share_button)

        # Let's now turn it back off
        share_button_form_data['has_email_share_button'] = False
        response = self.client.post(reverse("edit_petition", args=[p.id]),
                                    share_button_form_data)
        self.assertEqual(response.status_code, 200)
        p.refresh_from_db()
        self.assertFalse(p.has_email_share_button)

        # For a user owned petition
        p = Petition.objects.create(title="My petition 2", user=julia)

        # By default, a new petition has no share button
        self.assertFalse(p.has_email_share_button)

        # Now let's enable the email share button
        share_button_form_data['has_email_share_button'] = True
        response = self.client.post(reverse("edit_petition", args=[p.id]), share_button_form_data)
        self.assertEqual(response.status_code, 200)
        self.assertTemplateUsed(response, "petition/edit_petition.html")
        p.refresh_from_db()
        self.assertEquals(response.context['social_network_form'].is_valid(), True)
        self.assertEquals(response.context['social_network_form'].is_bound, True)
        self.assertEquals(response.context['content_form_submitted'], False)
        self.assertEquals(response.context['email_form_submitted'], False)
        self.assertEquals(response.context['social_network_form_submitted'], True)
        self.assertEquals(response.context['newsletter_form_submitted'], False)
        self.assertTrue(p.has_email_share_button)

        # Let's now turn it back off
        share_button_form_data['has_email_share_button'] = False
        response = self.client.post(reverse("edit_petition", args=[p.id]), share_button_form_data)
        self.assertEqual(response.status_code, 200)
        p.refresh_from_db()
        self.assertFalse(p.has_email_share_button)
```

This tests that when creating a Petition, the "has_email_share_button" is by default "off" and then tries to enable it and switch it back off using the "edit_petition" view.
This is done twice: once for an organization owned Petition and once for a user owned Petition.

I hope you are well and I wish you a very happy festive season :)

**Comment by rsk2** [comment]
@fallen Wish you a very happy festive season too!

**Comment by rsk2** [comment]
Hi @fallen, hope all is well. The PR is ready for review and please take a look whenever you get the chance.

**Review by fallen — APPROVED** [review]
Everything looks good to me, thanks a lot and sorry for the review delay!

**Pull request merged**